spec.html

```diff
@@ -131,112 +131,109 @@ <h1>CreateSyntheticModule ( _exportNames_, _evaluationSteps_, _realm_, _hostDefi
+          1. Set the VariableEnvironment of _moduleContext_ to _module_.[[Environment]].
+          1. Set the LexicalEnvironment of _moduleContext_ to _module_.[[Environment]].
+          1. Push _moduleContext_ on to the execution context stack; _moduleContext_ is now the running execution context.
+          1. Let _result_ be the result of performing _module_.[[EvaluationSteps]](_module_).
+          1. Suspend _moduleContext_ and remove it from the execution context stack.
           1. Resume the context that is now on the top of the execution context stack as the running execution context.
           1. Return Completion(_result_).
         </emu-alg>
       </emu-clause>
+
+      <emu-clause id="sec-synthetic-module-record-set-synthetic-export">
+        <h1>SetSyntheticExport ( _name_, _value_ )</h1>
+
+        <p>The SetSyntheticExport concrete method of a Synthetic Module Record _module_ takes argument _name_ (a String) and _value_.</p>
+        <p>It performs the following steps when called:</p>
+
+        <emu-alg>
+          1. Return ? _module_.[[Environment]].SetMutableBinding(_name_, _value_, *true*).
+        </emu-alg>
+      </emu-clause>
     </emu-clause>
   </emu-clause>
 
   <emu-clause id="sec-create-default-export-synthetic-module" aoid="CreateDefaultExportSyntheticModule">
-    <h1>CreateDefaultExportSyntheticModule ( _defaultExport_ )</h1>
-    <p>The CreateDefaultExportSyntheticModule abstract operation creates a Synthetic Module Record whose default export is _defaultExport_. It performs the following steps:</p>
+    <h1>CreateDefaultExportSyntheticModule ( _defaultExport_ , _realm_ , _hostDefined_ )</h1>
+    <p>The CreateDefaultExportSyntheticModule abstract operation takes arguments _defaultExport_, _realm_ (a Realm), and _hostDefined_.</p>
+    <p>It performs the following steps when called:</p>
     <emu-alg>
-      1. Return CreateSyntheticModule(&laquo; *"default"* &raquo;, the following steps, the current Realm, _defaultExport_) with the following steps given _module_ as an argument:</p>
-        1. SetSyntheticModuleExport(_module_, *"default"*, _module_.[[HostDefined]]).</li>
+      1. Let _closure_ be the a Abstract Closure with parameters (_module_) that captures _defaultExport_ and performs the following steps when called:
+        1. Return ? _module_.SetSyntheticExport(*"default"*, _defaultExport_).
+      1. Return CreateSyntheticModule(&laquo; *"default"* &raquo;, _closure_, _realm_)
     </emu-alg>
-    <!-- TODO: Could we apply the new spec closure concept here rather than painstakingly passing _defaultExport_ through [[HostDefined]]? -->
   </emu-clause>
 
   <emu-clause id="sec-parse-json-module" aoid="ParseJSONModule">
-    <h1>ParseJSONModule ( _source_ )</h1>
-    <p>The abstract operation ParseJSONModule takes a single argument _source_ which is a String representing the contents of a module.</p>
+    <h1>ParseJSONModule ( _sourceText_ , _realm_ , _hostDefined_ )</h1>
```

Comment:
The call site doesn't specify arguments for either version of the signature.

> If _type_ is *"json"*, then this algorithm must either invoke ParseJSONModule and return the resulting Completion Record, or throw an exception.

Is that intentional?

Reply:
which call site?

Reply:
The reference in "Runtime Semantics: HostResolveImportedModule".

Reply:
ah, I think so. the steps are host defined, so we mandate that they call it, not what they call it with.

Reply:
Ah, thanks! Ambiguity for the host-defined algorithms seems appropriate--I think what tripped me up is the difference in how we're acknowledging input and output for that operation. We don't mandate what implementations call the abstract operation with, but we do mandate what they do with the result. In a way, the function seems like the boundary for the normative text, so it's hard for me to classify the quoted statement as strictly normative or strictly informative.

Seems like I'm alone in that boat, though, so I'm happy to let it ride :)